### PR TITLE
[bug] MatrixType bug fix: Fix error with nested StructType and MatrixType

### PR DESCRIPTION
**Pull request by jim19930609**
Issue: https://github.com/taichi-dev/taichi/issues/5819

### Brief Summary
1. Modified `Matrix::fill()` to broadcast `val` into VectorType if `ndim == 1`, and to MatrixType if `ndim == 2`
2. Modified `Struct::fill()` to apply `matrix_op.fill()` in case of `Expr` with TensorType

### CODE REVIEW

File: python/taichi/lang/struct.py [hunk 1]
```diff
@@ -222,10 +222,20 @@ def fill(self, val):
         Args:
             val (Union[int, float]): Value to fill.
         """
-        def assign_renamed(x, y):
-            return ops.assign(x, y)
 
-        return self._element_wise_writeback_binary(assign_renamed, val)
+        bcast_val = self._broadcast_copy(val)
```
Comment (strongoier) [inline python/taichi/lang/struct.py:226]: Is it possible to simply use `val`? It seems that `_element_wise_binary` will also broadcast.

File: python/taichi/lang/struct.py [hunk 2]
```diff
@@ -222,10 +222,20 @@ def fill(self, val):
...
+            if isinstance(v, impl.Expr) and v.ptr.is_tensor():
+                from taichi.lang import matrix_ops  # pylint: disable=C0415
+                matrix_ops.fill(v, val)
+                entries[k] = v
+            elif isinstance(v, (Struct, Matrix)):
+                entries[k] = v._element_wise_binary(ops.assign,
+                                                    bcast_val.entries[k])
+            else:
+                entries[k] = ops.assign(v, bcast_val.entries[k])
+        return Struct(entries)
```
Comment (strongoier) [inline python/taichi/lang/struct.py:238]: The original `fill()` seems a bit problematic. `fill()` should happen in-place and no return value is needed.

File: taichi/transforms/lower_ast.cpp [hunk 3]
```diff
@@ -397,6 +397,7 @@ class LowerAST : public IRVisitor {
     auto fctx = make_flatten_ctx();
     flatten_rvalue(expr, &fctx);
     if (dest.is<IdExpression>()) {
+      flatten_lvalue(dest, &fctx);
```
Comment (strongoier) [inline taichi/transforms/lower_ast.cpp:400]: nit: move this outside as all branches do the same thing

File: taichi/transforms/lower_ast.cpp [hunk 4]
```diff
@@ -397,6 +397,7 @@ class LowerAST : public IRVisitor {
...
       fctx.push_back<LocalStoreStmt>(
           assign->parent->lookup_var(assign->lhs.cast<IdExpression>()->id),
           expr->stmt);
```
Comment (strongoier) [inline taichi/transforms/lower_ast.cpp:403]:
```suggestion
      fctx.push_back<LocalStoreStmt>(dest->stmt, expr->stmt);
```
After `flatten_lvalue` we can simply do this.

File: python/taichi/lang/struct.py [hunk 5]
```diff
@@ -222,10 +222,16 @@ def fill(self, val):
         Args:
             val (Union[int, float]): Value to fill.
         """
-        def assign_renamed(x, y):
-            return ops.assign(x, y)
-
-        return self._element_wise_writeback_binary(assign_renamed, val)
+        entries = {}
```
Comment (strongoier) [inline python/taichi/lang/struct.py:225]: Is `entries` also unnecessary?

Comment (jim19930609) [reply python/taichi/lang/struct.py:225]: good point, thanks!

**Pull request merged**